apps/src/music/blockly/MusicBlocklyWorkspace.js

```diff
@@ -20,17 +20,25 @@ import {
 } from './extensions';
 import experiments from '@cdo/apps/util/experiments';
 import {GeneratorHelpersSimple2} from './blocks/simple2';
+import ProjectManager from '@cdo/apps/labs/ProjectManager';
+import {S3SourcesStore, LocalSourcesStore} from '@cdo/apps/labs/SourcesStore';
+import {
+  S3ChannelsStore,
+  LocalChannelsStore
+} from '@cdo/apps/labs/ChannelsStore';
 import FieldChord from './FieldChord';
 
 /**
  * Wraps the Blockly workspace for Music Lab. Provides functions to setup the
- * workspace view, execute code, and save/load projects from local storage.
+ * workspace view, execute code, and save/load projects.
  */
 export default class MusicBlocklyWorkspace {
-  constructor() {
+  constructor(channelId) {
     this.codeHooks = {};
     this.compiledEvents = null;
     this.lastExecutedEvents = null;
+    this.channel = {id: channelId};
+    this.projectManager = this.getProjectManager(channelId);
```

Comment:
Rather than having information about the projectManager and channelId getting used within MusicBlocklyWorkspace, would it be possible to split these out and pass an interface into MusicBlocklyWorkspace that is used to trigger save and load actions?

Reasoning: Given that this class handles a lot of the logic required to run the code, it feels strange that it also knows how to use a channelId and knows how to determine whether sources are saved locally or remotely.

Reply:
The idea here is the `ProjectManager` is the interface that labs can interact with. Per local vs remote, I think it is ok for Music Lab to handle that--we will want labs to be able to decide how they save code (ex. db or s3), so it makes sense that they set up their own `ProjectManager`. Per the channel id question, right now we are getting it sent via the server, because we are integrating with our existing models in that respect. So I'm not sure how we avoid some part of MusicLab knowing about channel ids. A potential follow-up would be to have a new api that gets the channel id for a given level/project, and then we could extract the channel id fetch process out to the projects layer.
I might be misunderstanding what you mean here though, what would it look like to pass an interface in?

Reply:
Another thing to note is parts of the channel are relevant to the lab--for example the name of the project.

Reply:
I did some reworking of this pattern--there is now a `ProjectManagerFactory` that extracts most of setting up the project manager. I realized we needed to pass around `appOptions` for now in order to get the projectType, so it made sense to move this logic elsewhere. We are now passing in `appOptions` instead of `channelId` though. As we think about hot-reloading this pattern may change. For now I'm trying to not make too many changes at once.